In [3]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.decomposition import PCA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

# 1. Загрузить датасет WINE.

In [4]:
wine = load_wine(as_frame=True)
X = wine.data[['magnesium', 'flavanoids', 'proanthocyanins']]
y = wine.target

# 2. Подготовить выборки: обучающую 70% и тестовую 30%.

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [21]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# 3. Построить указанные в задании классификаторы.

In [22]:
knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train_scaled, y_train)
y_pred_knn = knn.predict(X_test_scaled)
accuracy_knn = accuracy_score(y_test, y_pred_knn)
accuracy_knn

0.8333333333333334

In [23]:
nb = GaussianNB()
nb.fit(X_train_scaled, y_train)
y_pred_nb = nb.predict(X_test_scaled)
accuracy_nb = accuracy_score(y_test, y_pred_nb)
accuracy_nb

0.7962962962962963

# 4. Провести анализ главных компонент с числом компонент =2.

In [24]:
pca = PCA(n_components=2)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

# 5. Получить оценку качества классификации в новом пространстве признаков с использованием указанных классификаторов.

In [25]:
knn_pca = KNeighborsClassifier(n_neighbors=3)
knn_pca.fit(X_train_pca, y_train)
y_pred_knn_pca = knn_pca.predict(X_test_pca)
accuracy_knn_pca = accuracy_score(y_test, y_pred_knn_pca)
accuracy_knn_pca

0.7407407407407407

In [26]:
nb_pca = GaussianNB()
nb_pca.fit(X_train_pca, y_train)
y_pred_nb_pca = nb_pca.predict(X_test_pca)
accuracy_nb_pca = accuracy_score(y_test, y_pred_nb_pca)
accuracy_nb_pca

0.7407407407407407

# 6. Провести линейный дискриминантный анализ путем построения одного признака.

In [27]:
lda = LinearDiscriminantAnalysis(n_components=1)
X_train_lda = lda.fit_transform(X_train_scaled, y_train)
X_test_lda = lda.transform(X_test_scaled)


# 7. Получить оценку качества классификации для построенного признака с использованием указанных классификаторов.

In [28]:
knn_lda = KNeighborsClassifier(n_neighbors=3)
knn_lda.fit(X_train_lda, y_train)
y_pred_knn_lda = knn_lda.predict(X_test_lda)
accuracy_knn_lda = accuracy_score(y_test, y_pred_knn_lda)
accuracy_knn_lda

0.8333333333333334

In [29]:
nb_lda = GaussianNB()
nb_lda.fit(X_train_lda, y_train)
y_pred_nb_lda = nb_lda.predict(X_test_lda)
accuracy_nb_lda = accuracy_score(y_test, y_pred_nb_lda)
accuracy_nb_lda

0.8703703703703703

# 8. Сравнить результаты классификации в п5 и п 7 с лучшей парой признаков и классификатором из Задания 4.

Лучшие пары из задания 4:

Accuracy of KNeighbors for [flavanoids, proanthocyanins]: 0.87

Accuracy of KNeighbors for [magnesium, flavanoids]: 0.87

# 9. Сделать вывод.

In [30]:
print(f"KNN Accuracy (Original): {accuracy_knn}")
print(f"NB Accuracy (Original): {accuracy_nb}")
print(f"KNN Accuracy (PCA): {accuracy_knn_pca}")
print(f"NB Accuracy (PCA): {accuracy_nb_pca}")
print(f"KNN Accuracy (LDA): {accuracy_knn_lda}")
print(f"NB Accuracy (LDA): {accuracy_nb_lda}")

KNN Accuracy (Original): 0.8333333333333334
NB Accuracy (Original): 0.7962962962962963
KNN Accuracy (PCA): 0.7407407407407407
NB Accuracy (PCA): 0.7407407407407407
KNN Accuracy (LDA): 0.8333333333333334
NB Accuracy (LDA): 0.8703703703703703


Методы снижения размерности (PCA и LDA) показали различные результаты: LDA продемонстрировал лучшую точность (87.04%) по сравнению с PCA. PCA показал немного худшие результаты.

LDA показал одинаковую или худшую точность для обеих моделей по сравнению с оригинальными признаками.

Использование методов снижения размерности в целом привело к улучшению результатов классификации только в случае с LDA.

LDA оказался более эффективным методом для данной задачи, что может быть связано с тем, что он учитывает информацию о классах при построении нового признака.

Разница в результатах между PCA и LDA подчеркивает важность выбора метода снижения размерности. LDA оказался более эффективным, что связано с тем, что он адаптирован к задаче классификации, тогда как PCA — более общий метод.